### ISSUE TITLE
Test filters passed into NUnit programmatically are not escaped and cause invalid XML characters to never match

### ISSUE THREAD

**Issue by MatthewBeardmore** [nunit/nunit#3063]
When NUnit VS adapter passes a test filter generated by VS into NUnit, the filter contains an invalid XML character (but valid UTF16 character) - for example, '\x01'. When read into TestFilter.LoadXML, it throws an exception and the test can never be run. Even if the XML character is escaped in the VS test adapter, it would still fail in FullNameFilter, as that expects the unescaped UTF16 character at that time.

**Comment by jnm2** [comment]
> When NUnit VS adapter passes a test filter generated by VS into NUnit, the filter contains an invalid XML character (but valid UTF16 character) - for example, '\x01'.

This should be fixed in the adapter. No good can come of constructing and passing invalid XML between systems, and we don't want to hide the symptoms of a more fundamental problem. my preference is for the framework to fail fast and loudly until the adapter sends it well-formed XML.

**Comment by jnm2** [comment]
More info:

The root of the problem starts in the framework when test names contain control characters. The framework is failing to remove the control characters from the test name values it encodes in XML. It must remove them because XML completely bans them and does not even permit them to be escaped using XML entities.

In practice, if the binary data is important, folks use base64 for this. Since the name is just for display purposes, I suggest we strip out the control characters by replacing them with our standard C-style display format: `\0`, \u0001`, etc.

**Comment by jnm2** [comment]
I'm slowly catching up to where you all are!
I believe this problem should be solved in three places:

1. The framework should prevent test names from containing control characters by replacing them with C-style escaping for display purposes. There should be no such thing as `ITest.FullName` containing control characters.
   ➡ All test names can be meaningfully seen and entered, and both XML output and filtering automatically work for free. Each runner doesn't have to reinvent ways to allow users to see and enter control characters in test names.
   ➡ Tests with parameters `"\u0001"` (control char) and `"\\u0001"` (literal text `\u0001`) would both be given the same name, literally the text `Foo("\u0001")`. For default name generation, I think this is ideal.

2. No need for filters to ever represent control characters, because no test names have control characters. The engine's filter builder should throw an exception if you pass it values that contain control characters. (It's currently also buggy because it doesn't handle UTF16 surrogate pairs. Both issues at https://github.com/nunit/nunit-console/issues/491).

3. VSTest's source-based discovery guesses at how the framework generates test case names. If it passes the adapter a guessed test name that contains a control character, the adapter will need to apply the framework's logic to properly format them. This might be what's happening at https://github.com/nunit/nunit3-vs-adapter/issues/484.

**Comment by OsirisTerje** [comment]
@jnm2  I do agree, but will this be done ?

**Comment by ChrisMaddock** [comment]
@nunit/framework-team Can I suggest this as a candidate for 4.0? It's more of a bug fix than a breaking change, but would be good to get resolved. I plan to tackle the engine part of this issue soon. 

Mentioning it as someone else has just been bit by it today: https://stackoverflow.com/q/66801413/1768779

**Comment by stevenaw** [comment]
I agree this would be great to fix. I would take this on myself, but unfortunately my development environment appears to have broken several weeks back (coinciding with after a windows update).

**Comment by Konamiman** [comment]
I'm experiencing this issue with Visual Studio 2022, NUnit 3.13.3, and NUnit3TestAdapter 4.2.1. A test like this won't run:

```c#
[TestCase("\u0001 abc")]
public void TestSomething(string someParam) {
    //Test something!
}
```

As a workaround I have modified my tests like this:

```c#
[TestCase("UNICODE_ONE abc")]
public void TestSomething(string someParam) {
    someParam = someParam.Replace("UNICODE_ONE", "\u0001");
    //Test something!
}
```
